system/jlib/jqueue.hpp

```diff
@@ -0,0 +1,387 @@
+            {
+                if (--numSpins != 0) // likely
+                {
+                    curState = state.load(std::memory_order_acquire);
+                    continue;
+                }
+                numSpins = initialSpinsBeforeWait;
+
+                //The list is currently full, increment the number of writers waiting.
+                //This can never overflow...
+                const state_t nextState = curState + ((state_t)1 << writerShift);
+                if (state.compare_exchange_weak(curState, nextState, std::memory_order_relaxed))
+                {
+                    if (aborted.load(std::memory_order_acquire))
+                        return false;
+                    writers.wait();
+                    if (aborted.load(std::memory_order_acquire))
+                        return false;
+                    curState = state.load(std::memory_order_acquire);
+                }
+            }
+            else
+            {
+                //Increment the number of items (which can never overflow), and possibly decrease readers
+                state_t nextState = (curState + 1);
+
+                //If a reader is waiting then decrement the count, and signal later..
+                //Note, this test is a constant folded
+                if (readerBits == 1)
+                {
+                    //More efficient to perform an unconditional mask
+                    nextState &= ~readerMask;
+                }
+                else
+                {
+                    if ((curState & readerMask) != 0)
+                        nextState -= (1 << readerShift);
+                }
+                if (state.compare_exchange_weak(curState, nextState, std::memory_order_relaxed))
+                {
+                    unsigned slotMask = (fixedSlotBits ? fixedSlotMask : dynamicSlotMask);
+                    unsigned curDequeueSeq = (curState >> dequeueShift); // No need to mask since the top field
+                    unsigned curEnqueueSeq = (curDequeueSeq + curCount) & sequenceMask;
+                    unsigned filledSeq = (curEnqueueSeq + 1) & sequenceMask;
+                    unsigned curEnqueueSlot = curEnqueueSeq & slotMask;
+                    BufferElement & cur = values[curEnqueueSlot];
+
+                    //MORE: Another producer has been interrupted while writing to the same slot
+                    //or the consumer has not yet read from the slot.
+                    //spin until that has been consumed.
+                    unsigned spins = 0;
+                    while (cur.sequence.load(std::memory_order_acquire) != curEnqueueSeq)
+                    {
+                        if (slotUnavailableSpins != 0 && ++spins == slotUnavailableSpins)
+                        {
+                            ThreadYield();
+                            spins = 0;
+                        }
+                        else
+                            spinPause();
+                    }
```

Comment:
It's recommended  to pause each iteration (we should also add this call to the spin loops) - it improves performance on systems with hyper-threading.